cmd/gardener-controller-manager/app/gardener_controller_manager.go

```diff
@@ -302,49 +303,47 @@ func NewGardener(config *componentconfig.ControllerManagerConfiguration) (*Garde
 }
 
 // Run runs the Gardener. This should never exit.
-func (g *Gardener) Run(stopCh chan struct{}) error {
+func (g *Gardener) Run(ctx context.Context, cancel context.CancelFunc) error {
+	leaderElectionCtx, leaderElectionCancel := context.WithCancel(context.Background())
+
 	// Prepare a reusable run function.
-	run := func(stop <-chan struct{}) {
-		go g.startControllers(stopCh)
-		<-stop
-		select {
-		case <-stopCh:
-			// can only happen if stopCh is already closed because it's never written to it
-		default:
-			close(stopCh)
-		}
+	run := func(ctx context.Context) {
+		g.startControllers(ctx)
 	}
 
 	// Start HTTP server
-	go server.Serve(g.K8sGardenClient, g.K8sGardenInformers, g.Config.Server, stopCh)
+	go server.Serve(ctx, g.K8sGardenClient, g.K8sGardenInformers, g.Config.Server)
 	handlers.UpdateHealth(true)
 
 	// If leader election is enabled, run via LeaderElector until done and exit.
 	if g.LeaderElection != nil {
 		g.LeaderElection.Callbacks = leaderelection.LeaderCallbacks{
-			OnStartedLeading: func(stop <-chan struct{}) {
+			OnStartedLeading: func(_ context.Context) {
 				g.Logger.Info("Acquired leadership, starting controllers.")
-				run(stop)
+				run(ctx)
+				<-ctx.Done()
+				leaderElectionCancel()
 			},
 			OnStoppedLeading: func() {
-				g.Logger.Info("Lost leadership, cleaning up.")
-				close(stopCh)
+				g.Logger.Info("Lost leadership, terminating.")
+				cancel()
 			},
 		}
 		leaderElector, err := leaderelection.NewLeaderElector(*g.LeaderElection)
 		if err != nil {
 			return fmt.Errorf("couldn't create leader elector: %v", err)
 		}
-		leaderElector.Run()
-		return fmt.Errorf("lost lease")
+		leaderElector.Run(leaderElectionCtx)
+		return nil
 	}
 
 	// Leader election is disabled, thus run directly until done.
-	run(stopCh)
+	go run(ctx)
```

Comment:
What's the reason we execute the `run` method in a Go routine? It causes the method to exit not gracefully any more.